### ISSUE TITLE
Allow variables as type annotations

### ISSUE THREAD

**Issue by Ovsyanka83** [microsoft/pyright#5054]
**Is your feature request related to a problem? Please describe.**
We commonly use the following pattern in enums:
```python
MyEnum = Enum("MyEnum", {key: key for key in SOME_MAPPING.keys()}, type=StrEnum)
```

However, whenever we try to use it in a type annotation (such as `list[MyEnum]` in arbitrary code or `Literal[MyEnum.some_value]` in pydantic), we get an error:
```
Illegal type annotation: variable not allowed unless it is a type alias (reportGeneralTypeIssues)
```
**Describe the solution you'd like**
Please, help me understand whether there is some method to do this (outside of "IF TYPE_CHECKING" and setting MyEnum to `Any`) or add some feature to allow ignoring this error globally. Or maybe even restructure this feature in a way that allows us to work with enums, if this is possible and ok with you :)

**Comment by erictraut** [comment]
If you are interested in static type checking, I recommend against using this way to define enums. The technique prevents the type checker from understanding the contents of the enumeration. At best, the type checker will know nothing about your enumeration and will not detect any errors when you use it incorrectly.

By comparison, mypy generates an error for this code as well, although it emits the error on the line where the `Enum` call is made.

If you _must_ use this technique for some reason, then using `if TYPE_CHECKING` is probably the best solution.

I considered the following other options:
1. "Allow variables as type annotations". That's not a viable option. Variables cannot be used as type annotations, and it's important that this is an error case.
2. Change the `Enum` special-case handling to ignore dictionary comprehensions in the functional form of `Enum` declarations. This would create a new `MyEnum` class with your code above, and `MyEnum` could then be used as a type annotation because it's a legitimate class. However, any attempt to access enum elements on this class (e.g. `MyEnum.foo`) would generate errors because the type checker doesn't know about them. I don't think you'd be satisfied with this solution.
3. Emit an error on the `Enum` call site. This would match mypy's behavior. It wouldn't eliminate the later "variable not allowed..." error, but it might make it clearer why you're seeing the later error. I don't think you'd be satisfied with this solution.
4. Change the `Enum` special-case handling to create a new `MyEnum` class that effectively inherits from `Any`. That would make it a legitimate class, but it would be lobotomized from the perspective of the type checker. No errors would be emitted when it is used. That would probably satisfy you, but it would open a "hole" in type checking that would not be detected even in strict mode. For that reason, I don't think this is a good option.
5. Leave the current behavior as is. You won't be satisfied with this option, but I think it's the best one on balance.
